src/queue.rs

```diff
@@ -0,0 +1,224 @@
+            if head.is_null() {
+                return None;
+            }
+
+            // Get the next pointer.
+            let next = unsafe { (*head).next.load(Ordering::Acquire) };
+
+            // Try to set the head to the next pointer.
+            if self
+                .head
+                .compare_exchange(head, next, Ordering::AcqRel, Ordering::Acquire)
+                .is_ok()
+            {
+                // If we just set the head to zero, set the tail to zero as well.
+                if next.is_null() {
+                    self.tail.store(ptr::null_mut(), Ordering::SeqCst);
+                }
+
+                // The head has been updated. Dequeue the old head.
+                let head = unsafe { NonNull::new_unchecked(head) };
+                if unsafe { head.as_ref().listener.dequeue() } {
+                    // The EventListener for this node has been dropped, free the node.
+                    unsafe { self.dealloc(head) };
+                } else {
+                    // The event is still in use, return the node.
+                    return Some(head);
+                }
+            }
+        }
+    }
+
+    /// Orphan a node in the queue.
+    ///
+    /// The return value is Some if the node was notified, and true is the notification
+    /// was an additional notification.
+    pub(crate) unsafe fn orphan(&self, node: NonNull<Node>) -> Option<bool> {
+        let (needs_drop, notified) = unsafe { node.as_ref().listener.orphan() };
+
+        // If we need to deallocate the node, do so.
+        if needs_drop {
+            unsafe { self.dealloc(node) };
+        }
+
+        notified
+    }
+
+    /// Allocate a new node for a listener.
+    fn alloc(&self, listener: Listener) -> NonNull<Node> {
+        // Try to install a cached node.
+        if !self.cache.occupied.swap(true, Ordering::Acquire) {
+            // We can now initialize the node.
+            let node_ptr = self.cache.node.get() as *mut Node;
+
+            unsafe {
+                // Initialize the node.
+                node_ptr.write(Node::new(listener));
+
+                // Return the node.
+                NonNull::new_unchecked(node_ptr)
+            }
+        } else {
+            // We failed to install a cached node, so we need to allocate a new
+            // one.
+            unsafe { NonNull::new_unchecked(Box::into_raw(Box::new(Node::new(listener)))) }
+        }
+    }
+
+    /// Deallocate a node.
+    unsafe fn dealloc(&self, node: NonNull<Node>) {
+        // Is this node the current cache node?
+        if ptr::eq(self.cache.node.get() as *const Node, node.as_ptr()) {
+            // We can now clear the cache.
+            unsafe { (self.cache.node.get() as *mut Node).drop_in_place() };
+            self.cache.occupied.store(false, Ordering::Release);
+        } else {
+            // We need to deallocate the node on the heap.
+            unsafe { Box::from_raw(node.as_ptr()) };
+        }
+    }
```

Comment:
these methods (`alloc`, `dealloc`, `orphan`) appear to only act on the `CachedNode` and imo should be moved into an `impl` block associated with that. (and if these methods are still necessary, should be replaced with forwarding methods)

Reply:
I've moved `alloc` and `dealloc`, but `orphan` is, semantically, an important operation on the `Queue` so I've kept it there.